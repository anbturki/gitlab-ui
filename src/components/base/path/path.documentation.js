import description from './path.md';
import examples from './examples';

export default {
  description,
  examples,
  followsDesignSystem: true,
};
