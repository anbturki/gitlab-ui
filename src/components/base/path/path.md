## Usage

Paths are horizontal process flows composed of a series of "stages".
Like tabs, paths let users focus in on specific content at each stage
whilst still keeping all the other stages in view. Only one stage can
be active at a given time.

### Implemetation

The component should be initialized with an array of data objects. By
default, the first item in the array will be selected. This can be
overridden by passing in an object with the selected property set to
true.

```js
items: [
  {
    title: 'First',
  },
  {
    title: 'Second',
    selected: true
  },
  ...
```

Once an item has been selected the `selected` event will be emitted.
The emitted event will include the entire object at the selected index.

### Additional information

This component supports various themes and is mobile responsive.
