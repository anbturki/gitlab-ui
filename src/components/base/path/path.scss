// Variables
$path-chevron-z-index: -1;
$path-chevron-top: px-to-rem(4.5px);
$path-chevron-right: px-to-rem(-11px);
$path-chevron-dimension: px-to-rem(23px);
$path-chevron-border-radius: px-to-rem(10px);
$path-button-right-padding: 1.25 * $grid-size;
$path-chevron-drop-shadow: drop-shadow(0 0 px-to-rem(1px) $blue-500);
$path-chevron-transformation: rotate(45deg) skew(18deg, 18deg);
$path-fade-left-gradient: linear-gradient(to left, $transparent, $white, $white, $white);
$path-fade-right-gradient: linear-gradient(to right, $transparent, $white, $white, $white);

// Mixins
@mixin gl-path-active-item-color($color) {
  @include gl-text-white;
  background-color: $color;

  &::after {
    background-color: $color;
  }
}

@mixin gl-path-chevron {
  content: '';
  @include gl-absolute;
  @include gl-bg-white;
  top: $path-chevron-top;
  right: $path-chevron-right;
  width: $path-chevron-dimension;
  z-index: $path-chevron-z-index;
  height: $path-chevron-dimension;
  transform: $path-chevron-transformation;
  border-top-left-radius: $path-chevron-border-radius;
  border-bottom-right-radius: $path-chevron-border-radius;
}

.gl-path-nav {
  @include gl-relative;
  @include gl-display-flex;

  .gl-path-nav-list {
    @include gl-p-1;
    @include gl-m-0;
    @include gl-relative;
    @include gl-list-style-none;
    @include gl-display-inline-flex;
    @include gl-overflow-hidden;
  }

  .gl-path-nav-list-item {
    @include gl-overflow-hidden;
    @include gl-min-w-fit-content;

    &:focus-within {
      filter: $path-chevron-drop-shadow;
    }
  }

  .gl-path-button {
    @include gl-pl-5;
    @include gl-py-3;
    @include gl-mr-4;
    @include gl-relative;
    @include gl-bg-white;
    @include gl-font-base;
    @include gl-z-index-0;
    @include gl-border-none;
    @include gl-text-gray-500;
    @include gl-font-weight-bold;
    @include gl-line-height-normal;
    @include gl-rounded-top-left-base;
    @include gl-rounded-bottom-left-base;
    padding-right: $path-button-right-padding;

    &:focus {
      @include gl-outline-none;
    }

    &::after {
      @include gl-path-chevron;
    }

    &:hover {
      @include gl-path-active-item-color($gray-100);
      @include gl-text-gray-900;
    }
  }

  .gl-path-fade {
    @include gl-w-8;
    @include gl-h-full;
    @include gl-bottom-0;
    @include gl-absolute;
    @include gl-z-index-2;
    @include gl-text-gray-500;

    &-left {
      @include gl-left-0;
      background-image: $path-fade-left-gradient;
    }

    &-right {
      @include gl-right-0;
      @include gl-text-right;
      background-image: $path-fade-right-gradient;
    }
  }

  .gl-clear-icon-button {
    @include gl-p-0;
    @include gl-text-gray-400;
    @include gl-h-full;
  }

  .gl-path-active-item-indigo,
  .gl-path-active-item-indigo:hover {
    @include gl-path-active-item-color($theme-indigo-700);
  }

  .gl-path-active-item-blue,
  .gl-path-active-item-blue:hover {
    @include gl-path-active-item-color($theme-blue-500);
  }

  .gl-path-active-item-light-blue,
  .gl-path-active-item-light-blue:hover {
    @include gl-path-active-item-color($theme-light-blue-500);
  }

  .gl-path-active-item-red,
  .gl-path-active-item-red:hover {
    @include gl-path-active-item-color($theme-red-500);
  }

  .gl-path-active-item-light-red,
  .gl-path-active-item-light-red:hover {
    @include gl-path-active-item-color($theme-light-red-500);
  }

  .gl-path-active-item-green,
  .gl-path-active-item-green:hover {
    @include gl-path-active-item-color($theme-green-500);
  }
}
