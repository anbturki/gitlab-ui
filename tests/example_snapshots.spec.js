import Vue from 'vue/dist/vue.esm';
import BootstrapVue from 'bootstrap-vue';
import path from 'path';
import fs from 'fs';

import { parseComponent } from 'vue-template-compiler';
import * as Babel from 'babel-standalone';

import puppeteer from 'puppeteer';
import { toMatchImageSnapshot } from 'jest-image-snapshot';

import * as gitlabComponents from '../index';
import { getResetAnimationsCSS } from './test_utils';

// Helper Function to get all files recursively in a directory
function getAllExampleFiles(dir) {
  return fs.readdirSync(dir).reduce((files, file) => {
    const fullPath = path.join(dir, file);

    const isDirectory = fs.statSync(fullPath).isDirectory();
    if (isDirectory) {
      return [...files, ...getAllExampleFiles(fullPath)];
    }
    return fullPath.endsWith('.example.vue') ? [...files, fullPath] : [...files];
  }, []);
}

function getExampleNameFromPath(examplePath) {
  return path
    .basename(examplePath, '.example.vue')
    .replace(/\./gi, '-')
    .replace(/_/gi, '-');
}

const basePath = path.resolve(__dirname, '../');
const allExamples = getAllExampleFiles(path.join(basePath, 'components'));

// We need to register all bootstrap and all GL UI Components so they are available when the example is mounted
Vue.use(BootstrapVue);
Object.keys(gitlabComponents).forEach(comp => {
  if (!comp.includes('Directive')) {
    Vue.component(comp, gitlabComponents[comp]);
  }
});

// We need to do Directives for now manually
Vue.directive('gl-tooltip', gitlabComponents.GlTooltipDirective);
Vue.directive('gl-modal-directive', gitlabComponents.GlModalDirective);
// Finished initialisation

describe('Snapshots for examples', () => {
  describe('Output Snapshots', () => {
    allExamples.map(async examplePath => {
      const relativePath = examplePath.replace(basePath, '');
      const exampleName = getExampleNameFromPath(relativePath);

      it(`${exampleName} should match the snapshot`, async () => {
        const exampleContent = fs.readFileSync(examplePath, 'utf8');
        const base = {
          name: '',
          template: '<div></div>',
        };

        // The loaded component is now parsed
        const parsed = parseComponent(exampleContent);

        let compiled = {};
        if (parsed.script) {
          const { code } = Babel.transform(parsed.script.content, {
            presets: ['es2015', 'stage-2'],
          });
          compiled = eval(`const exports = {};${code}`); // eslint-disable-line no-eval
        }

        compiled.template = parsed.template.content;

        // We have now a compiled (template + script) Vue component based on the example that we can mount
        const exampleControl = Object.assign({}, base, compiled);

        const Constructor = Vue.extend(exampleControl);
        const vm = new Constructor().$mount();

        expect(vm.$el).toMatchSnapshot();
      });
    });
  });

  describe('Image Snapshots', () => {
    let browser;

    beforeAll(async () => {
      // Extending Jest
      expect.extend({ toMatchImageSnapshot });

      // Starting the browser
      browser = await puppeteer.launch({
        headless: true,
        args: ['--no-sandbox', '--disable-setuid-sandbox'],
      });
    });

    afterAll(async () => {
      // Closing down our browser
      await browser.close();
    });

    allExamples.map(async examplePath => {
      const relativePath = examplePath.replace(basePath, '');
      const exampleName = getExampleNameFromPath(relativePath);

      it(`each example should match the image snapshot ${exampleName}`, async () => {
        const exampleUrl = `file://${path.resolve(
          __dirname,
          '../'
        )}/storybook/iframe.html?selectedKind=skeleton-loading&selectedStory=default&exampleName=${exampleName}`;

        try {
          const page = await browser.newPage();
          await page.goto(exampleUrl);

          // We wait until the example Box has loaded
          await page.waitForSelector('#exampleDisplay');

          // Removes the animation
          page.addStyleTag({
            content: getResetAnimationsCSS(),
          });

          const image = await page.screenshot();

          await expect(image).toMatchImageSnapshot({
            customSnapshotIdentifier: exampleName,
          });
        } catch (e) {
          console.error('Error ', e); // eslint-disable-line no-console
        }
      });
    });
  });
});
