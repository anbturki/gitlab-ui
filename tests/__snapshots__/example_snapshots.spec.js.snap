// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots for examples Output Snapshots button-basic should match the snapshot 1`] = `
<button
  class="btn btn-secondary"
  type="button"
>
  Test Button
</button>
`;

exports[`Snapshots for examples Output Snapshots button-link should match the snapshot 1`] = `
<a
  class="btn btn-secondary"
  href="#"
>
  Link Button
</a>
`;

exports[`Snapshots for examples Output Snapshots button-sizes should match the snapshot 1`] = `
<div>
  <button
    class="btn btn-secondary"
    type="button"
  >
    default
  </button>
   
  <button
    class="btn btn-secondary btn-lg"
    size="lg"
    type="button"
  >
    lg
  </button>
   
  <button
    class="btn btn-secondary btn-sm"
    size="sm"
    type="button"
  >
    sm
  </button>
</div>
`;

exports[`Snapshots for examples Output Snapshots button-variants should match the snapshot 1`] = `
<div>
  <button
    class="btn btn-primary"
    type="button"
    variant="primary"
  >
    primary
  </button>
   
  <button
    class="btn btn-secondary"
    type="button"
    variant="secondary"
  >
    secondary
  </button>
   
  <button
    class="btn btn-warning"
    type="button"
    variant="warning"
  >
    warning
  </button>
   
  <button
    class="btn btn-danger"
    type="button"
    variant="danger"
  >
    danger
  </button>
   
  <button
    class="btn btn-link"
    type="button"
    variant="link"
  >
    link
  </button>
</div>
`;

exports[`Snapshots for examples Output Snapshots link-basic should match the snapshot 1`] = `
<a
  class=""
  href="https://about.gitlab.com"
>
  Test Link
</a>
`;

exports[`Snapshots for examples Output Snapshots loading-icon-basic should match the snapshot 1`] = `
<div
  class="loading-container text-center"
>
  <i
    aria-hidden="true"
    aria-label="Loading"
    class="fa fa-spinner fa-1x fa-spin"
  />
</div>
`;

exports[`Snapshots for examples Output Snapshots loading-icon-big should match the snapshot 1`] = `
<div
  class="loading-container text-center"
>
  <i
    aria-hidden="true"
    aria-label="Loading"
    class="fa fa-spinner fa-5x fa-spin"
  />
</div>
`;

exports[`Snapshots for examples Output Snapshots loading-icon-inline should match the snapshot 1`] = `
<div>
  <span
    class="loading-container text-center"
  >
    <i
      aria-hidden="true"
      aria-label="Loading"
      class="fa fa-spinner fa-1x fa-spin"
    />
  </span>
   Testing Inline Icon

</div>
`;

exports[`Snapshots for examples Output Snapshots modal-basic should match the snapshot 1`] = `
<div>
  <button
    type="button"
  >
    
    Open modal
  
  </button>
   
  <div>
    <!---->
  </div>
</div>
`;

exports[`Snapshots for examples Output Snapshots pagination-basic should match the snapshot 1`] = `
<ul
  aria-disabled="false"
  aria-label="Pagination"
  class="gl-pagination pagination b-pagination pagination-md"
  role="menubar"
>
  <!---->
  <li
    class="page-item"
    role="none presentation"
  >
    <a
      aria-label="Goto previous page"
      class="page-link"
      href="#"
      role="menuitem"
      tabindex="-1"
      target="_self"
    >
      <span
        aria-hidden="true"
      >
        ‹
      </span>
    </a>
  </li>
  <!---->
  <li
    class="page-item"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 1"
      aria-posinset="1"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      1
    </a>
  </li>
  <li
    class="page-item active"
    role="none presentation"
  >
    <a
      aria-checked="true"
      aria-label="Goto page 2"
      aria-posinset="2"
      aria-setsize="10"
      class="page-link btn-primary"
      href="#"
      role="menuitemradio"
      tabindex="0"
      target="_self"
    >
      2
    </a>
  </li>
  <li
    class="page-item"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 3"
      aria-posinset="3"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      3
    </a>
  </li>
  <li
    class="page-item d-none d-sm-flex"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 4"
      aria-posinset="4"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      4
    </a>
  </li>
  <li
    class="page-item d-none d-sm-flex"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 5"
      aria-posinset="5"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      5
    </a>
  </li>
  <li
    class="page-item d-none d-sm-flex"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 6"
      aria-posinset="6"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      6
    </a>
  </li>
  <li
    class="page-item d-none d-sm-flex"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 7"
      aria-posinset="7"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      7
    </a>
  </li>
  <li
    class="page-item d-none d-sm-flex"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 8"
      aria-posinset="8"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      8
    </a>
  </li>
  <li
    class="page-item d-none d-sm-flex"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 9"
      aria-posinset="9"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      9
    </a>
  </li>
  <li
    class="page-item d-none d-sm-flex"
    role="none presentation"
  >
    <a
      aria-checked="false"
      aria-label="Goto page 10"
      aria-posinset="10"
      aria-setsize="10"
      class="page-link"
      href="#"
      role="menuitemradio"
      tabindex="-1"
      target="_self"
    >
      10
    </a>
  </li>
  <!---->
  <li
    class="page-item"
    role="none presentation"
  >
    <a
      aria-label="Goto next page"
      class="page-link"
      href="#"
      role="menuitem"
      tabindex="-1"
      target="_self"
    >
      <span
        aria-hidden="true"
      >
        ›
      </span>
    </a>
  </li>
  <!---->
</ul>
`;

exports[`Snapshots for examples Output Snapshots progress-bar-basic should match the snapshot 1`] = `
<div
  class="progress"
  value="30"
>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="30"
    class="progress-bar"
    role="progressbar"
    style="width: 30%;"
  >
    <!---->
  </div>
</div>
`;

exports[`Snapshots for examples Output Snapshots skeleton-loading-basic should match the snapshot 1`] = `
<div
  class="animation-container"
>
  <div
    class="skeleton-line-1"
  />
  <div
    class="skeleton-line-2"
  />
  <div
    class="skeleton-line-3"
  />
</div>
`;

exports[`Snapshots for examples Output Snapshots skeleton-loading-one-line should match the snapshot 1`] = `
<div
  class="animation-container"
>
  <div
    class="skeleton-line-1"
  />
</div>
`;

exports[`Snapshots for examples Output Snapshots tooltip-component-default should match the snapshot 1`] = `
<div>
  <button
    class="btn btn-secondary"
    id="ref2"
    type="button"
  >
    Testbutton
  </button>
   
  <div
    aria-hidden="true"
    class="d-none"
    style="display: none;"
  >
    <div>
      
    some 
    
      <em>
        tooltip
      </em>
      
    text
  
    </div>
  </div>
</div>
`;

exports[`Snapshots for examples Output Snapshots tooltip-directive-bottom should match the snapshot 1`] = `
<button
  class="btn btn-secondary"
  data-original-title="some tooltip text"
  title=""
  type="button"
>
  
  Testbutton

</button>
`;

exports[`Snapshots for examples Output Snapshots tooltip-directive-default should match the snapshot 1`] = `
<button
  class="btn btn-secondary"
  data-original-title="some tooltip text"
  title=""
  type="button"
>
  
  Testbutton

</button>
`;

exports[`Snapshots for examples Output Snapshots tooltip-directive-left should match the snapshot 1`] = `
<div
  class="d-flex align-items-center justify-content-center p-3"
>
  <button
    class="btn btn-secondary"
    data-original-title="some tooltip text"
    title=""
    type="button"
  >
    
    Testbutton
  
  </button>
</div>
`;

exports[`Snapshots for examples Output Snapshots tooltip-directive-right should match the snapshot 1`] = `
<button
  class="btn btn-secondary"
  data-original-title="some tooltip text"
  title=""
  type="button"
>
  
  Testbutton

</button>
`;
